README
Howto setup the telepathy account:

mc-tool add whosthere/whatsapp "WhatsappAccount" string:phonenumber=[phonenumber incl. country code without leading+ or 00] string:password=[base64 encode]
mc-tool enable whosthere/whatsapp/account0
mc-tool auto-connect whosthere/whatsapp/account0 on

If empathy does not show any contacts, then try starting empathy from shell. If it does complain
about not being able to open the adressbook: rm .local/share/evolution/adressbook


debian/changelog
telepathy-whosthere (0.7ubuntu1) quantal; urgency=low

  * Fix ppa upload

 -- Matthias Gehre <M.Gehre@gmx.de>  Wed, 20 Mar 2013 18:25:27 -0500

telepathy-whosthere (0.7) quantal; urgency=low

  * Lower dependencies for backports
  * Add icon

 -- Matthias Gehre <M.Gehre@gmx.de>  Mon, 18 Mar 2013 14:53:35 -0500

telepathy-whosthere (0.6) quantal; urgency=low

  * Support for contacts sync
  * Bug fixes

 -- Matthias Gehre <M.Gehre@gmx.de>  Thu, 14 Mar 2013 21:17:25 -0500

telepathy-whosthere (0.5) quantal; urgency=low

  * Support for images/videos/contacts/audio/location

 -- Matthias Gehre <M.Gehre@gmx.de>  Mon, 11 Mar 2013 22:42:06 -0500

telepathy-whosthere (0.4) quantal; urgency=low

  * Updated depedencies

 -- Matthias Gehre <M.Gehre@gmx.de>  Sat, 09 Mar 2013 01:11:26 -0600

telepathy-whosthere (0.3) quantal; urgency=low

  * More contact list, presence, fixes

 -- Matthias Gehre <M.Gehre@gmx.de>  Sat, 09 Mar 2013 00:55:15 -0600

telepathy-whosthere (0.2) quantal; urgency=low

  * Added depedency on yowsup

 -- Matthias Gehre <M.Gehre@gmx.de>  Fri, 08 Mar 2013 18:10:58 -0600

telepathy-whosthere (0.1) quantal; urgency=low

  * Initial release

 -- Matthias Gehre <M.Gehre@gmx.de>  Fri, 08 Mar 2013 00:58:40 -0600


main.cpp
/*
 * Copyright (C) 2013 Matthias Gehre <gehre.matthias@gmail.com>
 *
 * This work is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This work is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this work; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <QCoreApplication>
#include <QDebug>

#include <TelepathyQt/BaseConnectionManager>
#include <TelepathyQt/Constants>
#include <TelepathyQt/Debug>
#include <TelepathyQt/Types>

#include "protocol.h"
#include "pythoninterface.h"

using namespace Tp;

int main(int argc, char *argv[])
{
    QCoreApplication a(argc, argv);
    
    Tp::registerTypes();
    Tp::enableDebug(true);
    Tp::enableWarnings(true);

    BaseProtocolPtr proto = BaseProtocol::create<Protocol>(
            QDBusConnection::sessionBus(),
            QLatin1String("whatsapp"));
    BaseConnectionManagerPtr cm = BaseConnectionManager::create(
            QDBusConnection::sessionBus(), QLatin1String("whosthere"));
    cm->addProtocol(proto);
    cm->registerObject();

    PythonInterface::initPython();
    return a.exec();
}


protocol.h
/*
 * Copyright (C) 2013 Matthias Gehre <gehre.matthias@gmail.com>
 *
 * This work is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This work is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this work; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

#pragma once

#include <TelepathyQt/BaseProtocol>

class Protocol : public Tp::BaseProtocol
{
    Q_OBJECT
    Q_DISABLE_COPY(Protocol)

public:
    Protocol(const QDBusConnection &dbusConnection, const QString &name);
    virtual ~Protocol();

    static Tp::SimpleStatusSpecMap getSimpleStatusSpecMap();
private:
    Tp::BaseConnectionPtr createConnection(const QVariantMap &parameters, Tp::DBusError *error);
    QString identifyAccount(const QVariantMap &parameters, Tp::DBusError *error);
    QString normalizeContact(const QString &contactId, Tp::DBusError *error);

    // Proto.I.Addressing
    QString normalizeVCardAddress(const QString &vCardField, const QString vCardAddress,
            Tp::DBusError *error);
    QString normalizeContactUri(const QString &uri, Tp::DBusError *error);

    Tp::BaseProtocolAddressingInterfacePtr addrIface;
    Tp::BaseProtocolAvatarsInterfacePtr avatarsIface;
    Tp::BaseProtocolPresenceInterfacePtr presenceIface;
};


connection.h
/*
 * Copyright (C) 2013 Matthias Gehre <gehre.matthias@gmail.com>
 *
 * This work is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This work is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this work; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

#pragma once

#include <tuple>
#include <QHash>
#include <TelepathyQt/BaseConnection>
#include <TelepathyQt/BaseChannel>
#include <boost/bimap.hpp>

#include "pythoninterface.h"

//There is no client with support for that
//#define USE_CAPTCHA_FOR_REGISTRATION

class YSConnection : public Tp::BaseConnection
{
    Q_OBJECT
    Q_DISABLE_COPY(YSConnection)
public:
    /* Implementation of BaseConnection */
    YSConnection( 	const QDBusConnection &  	dbusConnection,
                    const QString &  	cmName,
                    const QString &  	protocolName,
                    const QVariantMap &  	parameters);
    ~YSConnection();
    void connect(Tp::DBusError *error);
    QStringList inspectHandles(uint handleType, const Tp::UIntList& handles, Tp::DBusError *error);
    //QString uniqueName() const override;
    Tp::BaseChannelPtr createChannel(const QString& channelType, uint targetHandleType,
                                     uint targetHandle, Tp::DBusError *error);
    Tp::ContactAttributesMap getContactAttributes(const Tp::UIntList& handles,
                                                  const QStringList& interfaces,
                                                  Tp::DBusError *error);
    Tp::UIntList requestHandles(uint handleType, const QStringList& identifiers, Tp::DBusError* error);
    void messageAcknowledged(QString id);
    QString sendMessage(const QString& jid, const Tp::MessagePartList& message, uint flags, Tp::DBusError* error);
    uint setPresence(const QString& status, const QString& message, Tp::DBusError* error);
    Tp::ContactAttributesMap getContactListAttributes(const QStringList& interfaces, bool hold, Tp::DBusError* error);
    void requestSubscription(const Tp::UIntList& contacts, const QString& message, Tp::DBusError* error);


    void getContactsByVCardField(const QString& field, const QStringList& addresses,const QStringList& interfaces,
                                 Tp::AddressingNormalizationMap& addressingNormalizationMap,
                                 Tp::ContactAttributesMap& contactAttributesMap, Tp::DBusError* error);
    void getContactsByURI(const QStringList& URIs, const QStringList& interfaces,
                         Tp::AddressingNormalizationMap& addressingNormalizationMap,
                         Tp::ContactAttributesMap& contactAttributesMap, Tp::DBusError* error);
#ifdef USE_CAPTCHA_FOR_REGISTRATION
    /* Authentication */
    void getCaptchas(Tp::CaptchaInfoList& captchaInfoList, uint& numberRequired, QString& language, Tp::DBusError* error);
    QByteArray getCaptchaData(uint ID, const QString& mimeType, Tp::DBusError* error);
    void answerCaptchas(const Tp::CaptchaAnswers& answers, Tp::DBusError* error);
    void cancelCaptcha(uint reason, const QString& debugMessage, Tp::DBusError* error);
#endif
private slots:
    void on_yowsup_auth_success(QString phonenumber);
    void on_yowsup_auth_fail(QString mobilenumber, QString reason);
    void on_yowsup_status_dirty();
    void on_yowsup_message_received(QString msgId, QString jid, QString content, uint timestamp, bool wantsReceipt, QString pushName);
    void on_yowsup_disconnected(QString reason);
    void on_yowsup_receipt_messageSent(QString jid,QString msgId);
    void on_yowsup_receipt_messageDelivered(QString jid, QString msgId);
    void on_yowsup_presence_available(QString jid);
    void on_yowsup_presence_unavailable(QString jid);

    void on_yowsup_image_received(QString msgId,QString jid,QString preview,QString url,QString size,bool wantsReceipt);
    void on_yowsup_video_received(QString msgId,QString jid,QString preview,QString url,QString size,bool wantsReceipt);
    void on_yowsup_audio_received(QString msgId, QString jid, QString url, QString size, bool wantsReceipt);
    void on_yowsup_location_received(QString msgId,QString jid,QString name,QString preview,QString latitude,QString longitude,bool wantsReceipt);
    void on_yowsup_vcard_received(QString msgId,QString jid,QString name, QString data,bool wantsReceipt);

    void on_yowsup_group_imageReceived(QString msgId,QString gid,QString jid,QString preview,QString url,QString size,bool wantsReceipt);
    void on_yowsup_group_videoReceived(QString msgId,QString gid,QString jid,QString preview,QString url,QString size,bool wantsReceipt);
    void on_yowsup_group_audioReceived(QString msgId,QString gid,QString jid,QString url, QString size, bool wantsReceipt);
    void on_yowsup_group_locationReceived(QString msgId,QString gid,QString jid,QString name, QString preview,QString latitude, QString longitude, bool wantsReceipt);
    void on_yowsup_group_vcardReceived(QString msgId,QString gid,QString jid,QString name,QString data,bool wantsReceipt);
    void on_yowsup_group_messageReceived(QString msgId,QString gid,QString jid,QString content, int timestamp, bool wantsReceipt, QString pushName);

    void on_yowsup_notification_contactProfilePictureUpdated(QString jid, uint timestamp,QString msgId,QString pictureId, bool wantsReceipt);
    void on_yowsup_notification_contactProfilePictureRemoved(QString jid, uint timestamp,QString msgId, bool wantsReceipt);
    void on_yowsup_notification_groupParticipantAdded(QString gJid, QString jid, QString author, uint timestamp,QString msgId, bool wantsReceipt);
    void on_yowsup_notification_groupParticipantRemoved(QString gjid, QString jid, QString author, uint timestamp,QString msgId,bool wantsReceipt);
    void on_yowsup_notification_groupPictureUpdated(QString jid, QString author, uint timestamp, QString msgId, QString pictureId, bool wantsReceipt);
    void on_yowsup_notification_groupPictureRemoved(QString jid, QString author, uint timestamp, QString msgId, bool wantsReceipt);

    void on_yowsup_group_subjectReceived(QString msgId,QString fromAttribute,QString author,QString newSubject,uint timestamp,bool receiptRequested);
    void on_yowsup_profile_setStatusSuccess(QString jid, QString msgId);
    void on_yowsup_group_gotInfo(QString gid, QString jid, QString subject, QString subjectOwner, QString subjectT, QString creation);
private:
    void yowsup_messageReceived(QString msgId, QString jid, const Tp::MessagePartList& body, uint timestamp,
                                bool wantsReceipt, const QString &gid = QString());
    void yowsup_linked_data_received(const char* type, QString msgId, QString jid, QString preview,
                                     QString url,QString size,bool wantsReceipt,const QString& gid = QString());
    void yowsup_vcard_received(QString msgId,QString jid,QString name, QString data,bool wantsReceipt, QString gid = QString());
    void yowsup_location_received(QString msgId, QString jid,
                                                QString name, QString preview,
                                                QString latitude, QString longitude,
                                                bool wantsReceipt, QString gid = QString());
private:
    bool isValidContact(const QString& identifier);
    bool isContactId(const QString& jid);
    bool isGroupId(const QString& jid);
    Tp::HandleType getType(uint handle);
    Tp::HandleType getType(const QString& id);
    bool isValidHandle(uint handle);
    uint getHandle(const QString& id);
    QString getIdentifier(uint handle);
    uint ensureHandle(QString id);
    uint addGroup(const QString& gid);
    uint ensureGroup(QString gid);
    uint addContact(const QString& jid);
    uint addContacts(const QStringList& jid);
    uint ensureContact(QString jid);
    void setPresenceState(const QList<uint> handles, const QString& status);
    void setSubscriptionState(const QStringList& jid, const QList<uint> handles, uint state);
    QString generateUID();
    QString formatSize(QString size_);

    Tp::BaseConnectionRequestsInterfacePtr requestsIface;
    Tp::BaseConnectionContactsInterfacePtr contactsIface;
    Tp::BaseConnectionSimplePresenceInterfacePtr simplePresenceIface;
    Tp::BaseConnectionContactListInterfacePtr contactListIface;
    Tp::BaseConnectionAddressingInterfacePtr addressingIface;
#ifdef USE_CAPTCHA_FOR_REGISTRATION
    /* Only valid during registration */
    Tp::BaseChannelCaptchaAuthenticationInterfacePtr captchaIface;
#endif
    /* Maps ids to identifiers. handle "0" is never valid according to spec */
    boost::bimap<uint,QString> mHandles;
    /* Maps a contact handle to its subscription state */
    QHash<uint,uint> mContactsSubscription;

    /* increasing id for unique telepathy-ids */
    uint lastMessageId;
    uint selfHandle;

    PythonInterface* pythonInterface;

    QString mPhoneNumber;
    QByteArray mPassword;
    YowsupInterface yowsupInterface;
};


pythoninterface.h
/*
 * Copyright (C) 2013 Matthias Gehre <gehre.matthias@gmail.com>
 *
 * This work is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This work is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this work; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

#ifndef PYTHONINTERFACE_H
#define PYTHONINTERFACE_H

#include <thread>
#include <QObject>
#include <QString>
#include <boost/python.hpp>

/* Class which implements all signals emitted by yowsup */
class YowsupInterface : public QObject {
    Q_OBJECT
public:
    YowsupInterface(QObject *parent);
signals:
    void auth_success(QString mobilenumber);
    void auth_fail(QString mobilenumber, QString reason);
    void status_dirty();

    void message_received(QString msgId, QString jid, QString content, uint timestamp, bool wantsReceipt, QString pushName);
    void image_received(QString msgId,QString jid,QString preview,QString url,QString size,bool wantsReceipt);
    void video_received(QString msgId,QString jid,QString preview,QString url,QString size,bool wantsReceipt);
    void audio_received(QString msgId,QString jid,QString url,QString size,bool wantsReceipt);
    void location_received(QString msgId,QString jid,QString name,QString preview,QString latitude,QString longitude,bool wantsReceipt);
    void vcard_received(QString msgId,QString jid,QString name, QString data,bool wantsReceipt);

    void group_imageReceived(QString msgId,QString jid,QString author,QString preview,QString url,QString size,bool wantsReceipt);
    void group_videoReceived(QString msgId,QString jid,QString author,QString preview,QString url,QString size,bool wantsReceipt);
    void group_audioReceived(QString msgId,QString jid,QString author,QString url, QString size, bool wantsReceipt);
    void group_locationReceived(QString msgId,QString jid,QString author,QString name, QString preview,QString latitude, QString longitude, bool wantsReceipt);
    void group_vcardReceived(QString msgId,QString jid,QString author,QString name,QString data,bool wantsReceipt);
    void group_messageReceived(QString msgId,QString jid,QString author,QString content,int timestamp,bool wantsReceipt, QString pushName);

    void notification_contactProfilePictureUpdated(QString jid, uint timestamp,QString msgId,QString pictureId, bool wantsReceipt);
    void notification_contactProfilePictureRemoved(QString jid, uint timestamp,QString msgId, bool wantsReceipt);
    void notification_groupParticipantAdded(QString gJid, QString jid, QString author, uint timestamp,QString msgId, bool wantsReceipt);
    void notification_groupParticipantRemoved(QString gjid, QString jid, QString author, uint timestamp,QString msgId,bool wantsReceipt);
    void notification_groupPictureUpdated(QString jid, QString author, uint timestamp, QString msgId, QString pictureId, bool wantsReceipt);
    void notification_groupPictureRemoved(QString jid, QString author, uint timestamp, QString msgId, bool wantsReceipt);

    void disconnected(QString reason);
    void receipt_messageSent(QString jid,QString msgId);
    void receipt_messageDelivered(QString jid, QString msgId);
    void presence_updated(QString jid, QString lastSeen);
    void presence_available(QString jid);
    void presence_unavailable(QString jid);

    void group_subjectReceived(QString msgId,QString fromAttribute,QString author,QString newSubject,uint timestamp,bool receiptRequested);
    void profile_setStatusSuccess(QString jid, QString msgId);

    void group_setSubjectSuccess(QString jid);

    void group_gotInfo(QString jid,QString owner,QString subject,QString subjectOwner,QString subjectT,QString creation);


    void group_addParticipantsSuccess(QString jid, QString jids);
    void group_removeParticipantsSuccess(QString jid, QString jids);
    void group_createSuccess(QString jid);
    void group_createFail(QString errorCode);
    void group_endSuccess(QString jid);
    void group_gotPicture(QString jid, QString pictureId, QString filepath);
    void group_infoError(QString errorCode);
    void group_gotParticipants(QString jid, QString jids);
    void group_setPictureSuccess(QString jid, QString pictureId);
    void group_setPictureError(QString jid, QString errorCode);

    void profile_setPictureSuccess();
    void profile_setPictureError(QString errorCode);
    void receipt_visible(QString jid, QString msgId);
    void contact_gotProfilePictureId(QString jid, QString pictureId, QString filename);
    void contact_typing(QString jid);
    void contact_paused(QString jid);
    void contact_gotProfilePicture(QString jid, QString filename);
    void message_error(QString msgId,QString jid,QString errorCode);
    void ping(QString pingId);
    void pong();
};

class PythonInterface
{
public:
    PythonInterface(YowsupInterface* handler);
    ~PythonInterface();
    /* Call a python function on Yowsup's methodInterface */
    template<typename... T>
    boost::python::object call(const QString& method, const T&... args);
    /* Call a python function in our python wrapper */
    template<typename... T>
    boost::python::object call_intern(const char* method, const T&... args);
    /* Runs the thread reading from the connection to whatsapp. Signals
     *  will be dispatched from that thread.
     */
    void runReaderThread();
    /* One-time initialization */
    static void initPython();
private:
    static boost::python::object pModule;
    boost::python::object pConnectionManager;
    std::thread readerThread;
};

/* Class to hold ensure/release GIL lock */
class GILStateHolder
{
public:
    GILStateHolder();
    ~GILStateHolder();
private:
    PyGILState_STATE gstate;
};

#endif // PYTHONINTERFACE_H


YowsupInterface.py.h
/*
 * Copyright (C) 2013 Matthias Gehre <gehre.matthias@gmail.com>
 *
 * This work is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This work is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this work; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

const char* YowsupInterfacePy =
R"(
from Yowsup.connectionmanager import YowsupConnectionManager
from Yowsup.Contacts.contacts import WAContactsSyncRequest
from Yowsup.Registration.v2.coderequest import WACodeRequest as WACodeRequestV2
from Yowsup.Registration.v2.regrequest import WARegRequest as WARegRequestV2
from Yowsup.Common.debugger import Debugger
import signal
#Don't swallow SIGINT
signal.signal(signal.SIGINT, signal.SIG_DFL)
Debugger.enabled = False

def resultToString(result):
    out = []
    for k, v in result.items():
        if v is None:
            continue
        if type(v) is str:
            v = v.encode('utf-8')
        out.append("%s: %s" %(k, v))

    return "\n".join(out)

def createCallback(handler, sig):
    return lambda *args: getattr(handler,sig)(*args)

def init(handler):
    #!don't use any methods on handler before returning!
    connectionManager = YowsupConnectionManager()
    connectionManager.setAutoPong(True)
    signalsInterface = connectionManager.getSignalsInterface()
    for sig in signalsInterface.signals:
        signalsInterface.registerListener(sig, createCallback(handler, sig))
    return connectionManager

def call(connectionManager,methodName,*args):
    return connectionManager.getMethodsInterface().call(methodName, args)

def onSignal(i, *args):
    print 'YI: Got signal ' + i
    getattr(Emb, i)(*args)

def runThread(connectionManager):
    print 'In runThread'
    connectionManager.startReader()
    connectionManager.readerThread.join()

def syncContact(login, password, contact):
    wsync = WAContactsSyncRequest(login, password, (contact,))
    result = wsync.send()
    print resultToString(result)
    ret = result[u'c'][0][u'w']
    print "Returning ", ret
    return ret

def syncContacts(login, password, contacts):
    wsync = WAContactsSyncRequest(login, password, contacts.split(','))
    result = wsync.send()
    print resultToString(result)
    ret = []
    for i in result[u'c']:
        if i[u'w']:
            ret.append(i[u'n'].encode('utf-8'))
    print "Returning ", ret
    return ret

def code_request(self, countryCode, phoneNumber, identity, useText):
    print "code_request entered ", countryCode, phoneNumber, identity, "sms" if(useText) else "voice"
    we = WACodeRequestV2(countryCode, phoneNumber, identity, "sms" if(useText) else "voice")
    #result = we.send()
    #print resultToString(result)
    #return result['status']
    return u'123'

def code_register(self, countryCode, phoneNumber, identity, code):
    we = WARegRequestV2(countryCode, phoneNumber, code, identity)
    #result = we.send()
    #print resultToString(result)
    #return result['pw'] if result['pw'] else ''
    return u'ok'
)";


protocol.cpp
/*
 * Copyright (C) 2013 Matthias Gehre <gehre.matthias@gmail.com>
 *
 * This work is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This work is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this work; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include "protocol.h"
#include "connection.h"

#include <TelepathyQt/BaseConnection>
#include <TelepathyQt/Constants>
#include <TelepathyQt/RequestableChannelClassSpec>
#include <TelepathyQt/RequestableChannelClassSpecList>
#include <TelepathyQt/Types>

#include <QLatin1String>
#include <QVariantMap>

using namespace Tp;

SimpleStatusSpecMap Protocol::getSimpleStatusSpecMap()
{
    //Presence
    SimpleStatusSpec spAvailable;
    spAvailable.type = ConnectionPresenceTypeAvailable;
    spAvailable.maySetOnSelf = false;
    spAvailable.canHaveMessage = true;

    SimpleStatusSpec spOffline;
    spOffline.type = ConnectionPresenceTypeOffline;
    spOffline.maySetOnSelf = false;
    spOffline.canHaveMessage = false;

    SimpleStatusSpec spUnknown;
    spUnknown.type = ConnectionPresenceTypeUnknown;
    spUnknown.maySetOnSelf = false;
    spUnknown.canHaveMessage = false;

    SimpleStatusSpecMap specs;
    specs.insert(QLatin1String("available"), spAvailable);
    specs.insert(QLatin1String("offline"), spOffline);
    specs.insert(QLatin1String("unknown"), spUnknown);
    return specs;
}

Protocol::Protocol(const QDBusConnection &dbusConnection, const QString &name)
    : BaseProtocol(dbusConnection, name)
{
    setParameters(ProtocolParameterList()
        << ProtocolParameter(QLatin1String("account"),
                             QLatin1String("s"), ConnMgrParamFlagRequired)
        << ProtocolParameter(QLatin1String("password"),
                             QLatin1String("s"), ConnMgrParamFlagRequired | ConnMgrParamFlagSecret)
        /*<< ProtocolParameter(QLatin1String("uid"),
                             QLatin1String("s"), ConnMgrParamFlagRegister)*/);

    setRequestableChannelClasses(
            RequestableChannelClassSpecList() << RequestableChannelClassSpec::textChat()
                                              << RequestableChannelClassSpec::textChatroom());

    setEnglishName(QLatin1String("WhatsApp"));
    setIconName(QLatin1String("whosthere"));
    setVCardField(QLatin1String("tel"));
    //FIXME: setConnectionInterfaces(...);

    // callbacks
    setCreateConnectionCallback(memFun(this, &Protocol::createConnection));
    setIdentifyAccountCallback(memFun(this, &Protocol::identifyAccount));
    setNormalizeContactCallback(memFun(this, &Protocol::normalizeContact));

    // Adressing
    addrIface = BaseProtocolAddressingInterface::create();
    addrIface->setAddressableVCardFields(QStringList() << QLatin1String("tel"));
    addrIface->setAddressableUriSchemes(QStringList() << QLatin1String("tel"));
    addrIface->setNormalizeVCardAddressCallback(memFun(this, &Protocol::normalizeVCardAddress));
    addrIface->setNormalizeContactUriCallback(memFun(this, &Protocol::normalizeContactUri));
    plugInterface(AbstractProtocolInterfacePtr::dynamicCast(addrIface));

    // Avatars
    avatarsIface = BaseProtocolAvatarsInterface::create();
    avatarsIface->setAvatarDetails(AvatarSpec(QStringList() << QLatin1String("image/png"),
                16, 64, 32, 16, 64, 32, 1024));
    plugInterface(AbstractProtocolInterfacePtr::dynamicCast(avatarsIface));

    presenceIface = BaseProtocolPresenceInterface::create();
        presenceIface->setStatuses(PresenceSpecList(getSimpleStatusSpecMap()));
    plugInterface(AbstractProtocolInterfacePtr::dynamicCast(presenceIface));
}

Protocol::~Protocol()
{
}

BaseConnectionPtr Protocol::createConnection(const QVariantMap &parameters, Tp::DBusError *error) {

    if(!parameters.contains("account")) {
        error->set(TP_QT_ERROR_INVALID_ARGUMENT, QLatin1String("account is missing"));
        return BaseConnectionPtr();
    } else {
        return BaseConnection::create<YSConnection>( "whosthere", name().toLatin1(), parameters);
    }
}

QString Protocol::identifyAccount(const QVariantMap &parameters, Tp::DBusError *error)
{
    qDebug() << "Protocol::identifyAccount " << parameters;
    error->set(QLatin1String("IdentifyAccount.Error.Test"), QLatin1String(""));
    return QString();
}

QString Protocol::normalizeContact(const QString &contactId, Tp::DBusError *error)
{
    qDebug() << "Protocol::normalizeContact " << contactId;
    error->set(QLatin1String("NormalizeContact.Error.Test"), QLatin1String(""));
    return QString();
}

QString Protocol::normalizeVCardAddress(const QString &vcardField, const QString vcardAddress,
        Tp::DBusError *error)
{
    qDebug() << "Protocol::normalizeVCardAddress " << vcardField;
    error->set(QLatin1String("NormalizeVCardAddress.Error.Test"), QLatin1String(""));
    return QString();
}

QString Protocol::normalizeContactUri(const QString &uri, Tp::DBusError *error)
{
    qDebug() << "Protocol::normalizeContactUri " << uri;
    error->set(QLatin1String("NormalizeContactUri.Error.Test"), QLatin1String(""));
    return QString();
}
